lib/puppet/util/windows/taskscheduler2.rb

```diff
@@ -0,0 +1,707 @@
+#require 'puppet/util/windows/error'
+#require 'win32ole' # TODO should guard this on Windows only
+
+# The TaskScheduler2 class encapsulates taskscheduler settings and behavior using the v2 API
+# https://msdn.microsoft.com/en-us/library/windows/desktop/aa383600(v=vs.85).aspx
+
+# @api private
+class Puppet::Util::Windows::TaskScheduler2
+
+  # include Puppet::Util::Windows::String
+
+  # TODO May not need this if no COM
+  # require 'ffi'
+  # extend FFI::Library
+
+  # The error class raised if any task scheduler specific calls fail.
+  class Error < Puppet::Util::Windows::Error; end
+
+  # private
+
+  # # TODO needed?
+  # class << self
+  #   attr_accessor :com_initialized
+  # end
+
+  public
+
+  # The name of the root folder for tasks
+  ROOT_FOLDER = '\\'
+  
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa378137(v=vs.85).aspx
+  S_OK = 0
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383558(v=vs.85).aspx
+  TASK_ENUM_HIDDEN  = 0x1
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa380596(v=vs.85).aspx
+  TASK_ACTION_EXEC = 0
+  TASK_ACTION_COM_HANDLER = 5
+  TASK_ACTION_SEND_EMAIL = 6
+  TASK_ACTION_SHOW_MESSAGE = 7
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383557(v=vs.85).aspx
+  # Undocumented values
+  # Win7/2008 R2                       = 3
+  # Win8/Server 2012 R2 or Server 2016 = 4
+  # Windows 10                         = 6
+  TASK_COMPATIBILITY_AT = 0
+  TASK_COMPATIBILITY_V1 = 1
+  TASK_COMPATIBILITY_V2 = 2
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383617(v=vs.85).aspx
+  TASK_STATE_UNKNOWN  = 0
+  TASK_STATE_DISABLED = 1
+  TASK_STATE_QUEUED   = 2
+  TASK_STATE_READY    = 3
+  TASK_STATE_RUNNING  = 4
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa383915%28v=vs.85%29.aspx
+  TASK_TRIGGER_EVENT                 = 0
+  TASK_TRIGGER_TIME                  = 1
+  TASK_TRIGGER_DAILY                 = 2
+  TASK_TRIGGER_WEEKLY                = 3
+  TASK_TRIGGER_MONTHLY               = 4
+  TASK_TRIGGER_MONTHLYDOW            = 5
+  TASK_TRIGGER_IDLE                  = 6
+  TASK_TRIGGER_REGISTRATION          = 7
+  TASK_TRIGGER_BOOT                  = 8
+  TASK_TRIGGER_LOGON                 = 9
+  TASK_TRIGGER_SESSION_STATE_CHANGE  = 11
+
+  # https://msdn.microsoft.com/en-us/library/windows/desktop/aa382538%28v=vs.85%29.aspx
+  TASK_VALIDATE_ONLY                 = 0x1
+  TASK_CREATE                        = 0x2
+  TASK_UPDATE                        = 0x4
+  TASK_CREATE_OR_UPDATE              = 0x6
+  TASK_DISABLE                       = 0x8
+  TASK_DONT_ADD_PRINCIPAL_ACE        = 0x10
+  TASK_IGNORE_REGISTRATION_TRIGGERS  = 0x20
```

Comment:
Most of these are unused...

Reply:
So far yes.  As i don't know what features we will implement in the future, and it is "cheap" to copy them all in, I didn't see an issue.